File: client/src/voice/drivingReducer.ts
// Pure state reducer for the Driving screen. Exposed independently of
// the React hook so it can be tested without a render environment.
//
// States:
//   idle       → waiting for the user to tap start
//   recording  → mic is open; daemon is recording audio for OpenClaw infer STT
//   thinking   → mic closed; daemon running xAI chat on the transcript
//   ai         → daemon running OpenClaw infer TTS; phone is playing audio
//
// Events come from the UI (tap / silence) and from daemon control
// messages (stt.done / stt.error / reply.done / reply.error / tts.done
// / tts.error). The reducer returns the next state plus a list of
// side-effect intents that the hook performs outside the reducer.

export type DrivingState = 'idle' | 'recording' | 'thinking' | 'ai';


export interface DrivingContext {
  state: DrivingState;
  lastUserText: string;
  lastReplyText: string;
  // Reply text received before audio starts. Kept hidden until tts.start.
  pendingReplyText: string;
  // Render label source for the AI caption.
  liveReplyText: string;
  error: string | null;
}

export type DrivingReplayEvent =
  | { type: 'tap'; currentTurnTranscribing?: boolean }
  | { type: 'silence' }
  | { type: 'stt.done'; text: string }
  | { type: 'stt.error'; reason: string }
  | { type: 'reply.done'; text: string }
  | { type: 'reply.error'; reason: string }
  | { type: 'tts.start'; text?: string }
  | { type: 'tts.done' }
  | { type: 'tts.error'; reason: string };

export interface DrivingHydration {
  context: DrivingContext;
  armTts?: boolean;
}

export type DrivingEvent =
  | DrivingReplayEvent
  | { type: 'session.replay'; events: DrivingReplayEvent[]; hydration?: DrivingHydration }
  | { type: 'session.reset' };

export type DrivingSideEffect =
  | { kind: 'startMic' }
  | { kind: 'stopMic' }
  | { kind: 'cancelMic' }
  | { kind: 'armTts' }
  | { kind: 'stopTts' }
  | { kind: 'cancelReply' };

export interface Reduction {
  next: DrivingContext;
  side: DrivingSideEffect[];
}

export const initialContext: DrivingContext = {
  state: 'idle',
  lastUserText: '',
  lastReplyText: '',
  pendingReplyText: '',
  liveReplyText: '',
  error: null,
};


export function reduce(ctx: DrivingContext, event: DrivingEvent): Reduction {
  if (event.type === 'session.replay') return reduceSessionReplay(ctx, event);
  if (event.type === 'session.reset') return { next: { ...initialContext }, side: [] };

  switch (ctx.state) {
    case 'idle':
      if (event.type === 'tap') {
        return {
          next: { ...ctx, state: 'recording', error: null, pendingReplyText: '', liveReplyText: '' },
          side: [{ kind: 'startMic' }],
        };
      }
      if (event.type === 'tts.start') {
        return {
          next: {
            ...ctx,
            state: 'ai',
            ...(event.text ? { lastReplyText: event.text, liveReplyText: event.text } : {}),
            error: null,
          },
          side: [],
        };
      }
      return { next: ctx, side: [] };

    case 'recording':
      if (event.type === 'tap') {
        return {
          next: { ...ctx, state: 'thinking' },
          side: [{ kind: 'stopMic' }],
        };
      }
      if (event.type === 'stt.error') {
        return {
          next: { ...ctx, state: 'idle', error: event.reason },
          side: [{ kind: 'cancelMic' }],
        };
      }
      return { next: ctx, side: [] };

    case 'thinking':
      if (event.type === 'stt.done') {
        // Final transcript arrives while already in `thinking`. Don't
        // start anything new — daemon pipelines chat automatically.
        return { next: { ...ctx, lastUserText: event.text }, side: [] };
      }
      if (event.type === 'reply.done') {
        return {
          next: { ...ctx, pendingReplyText: event.text, liveReplyText: '' },
          side: [{ kind: 'armTts' }],
        };
      }
      if (event.type === 'tts.start') {
        const replyText = ctx.pendingReplyText || event.text || '';
        if (!replyText) return { next: ctx, side: [] };
        return {
          next: {
            ...ctx,
            state: 'ai',
            lastReplyText: replyText,
            liveReplyText: replyText,
            pendingReplyText: '',
          },
          side: [],
        };
      }
      if (event.type === 'reply.error' || event.type === 'stt.error') {
        return {
          next: { ...ctx, state: 'idle', error: event.reason, pendingReplyText: '', liveReplyText: '' },
          side: [],
        };
      }
      if (event.type === 'tts.done') {
        return {
          next: { ...ctx, state: 'idle', pendingReplyText: '', liveReplyText: '' },
          side: [],
        };
      }
      if (event.type === 'tts.error') {
        // Audio synthesis is non-fatal once the reply itself was generated:
        // promote the pending reply into the visible "last AI turn" so the
        // user can still read what was said and surface a soft audio error.
        return {
          next: {
            ...ctx,
            state: 'idle',
            error: event.reason,
            lastReplyText: ctx.pendingReplyText || ctx.lastReplyText,
            pendingReplyText: '',
            liveReplyText: '',
          },
          side: [],
        };
      }
      if (event.type === 'tap') {
        if (event.currentTurnTranscribing) {
          return { next: ctx, side: [] };
        }
        // Double-tap from thinking bails out of the turn after the
        // authoritative STT final has arrived and the reply turn is running.
        return {
          next: { ...ctx, state: 'idle', pendingReplyText: '', liveReplyText: '' },
          side: [{ kind: 'cancelReply' }],
        };
      }
      return { next: ctx, side: [] };

    case 'ai':
      if (event.type === 'tap' || event.type === 'silence') {
        return {
          next: { ...ctx, state: 'idle' },
          side: [{ kind: 'stopTts' }],
        };
      }
      if (event.type === 'tts.done') {
        return { next: { ...ctx, state: 'idle' }, side: [] };
      }
      if (event.type === 'tts.error') {
        return {
          next: { ...ctx, state: 'idle', error: event.reason, liveReplyText: '' },
          side: [],
        };
      }
      return { next: ctx, side: [] };
  }
}

function reduceSessionReplay(
  ctx: DrivingContext,
  event: Extract<DrivingEvent, { type: 'session.replay' }>,
): Reduction {
  let next = ctx;
  let side: DrivingSideEffect[] = [];
  const hasTerminalReplayEvent = event.events.some(isTerminalReplayEvent);
  for (const replayEvent of event.events) {
    const reduced = reduce(next, replayEvent);
    next = reduced.next;
    side.push(...reduced.side);
  }
  if (event.hydration) {
    const terminalReplayWins =
      hasTerminalReplayEvent && next.state === 'idle' && event.hydration.context.state !== 'idle';
    if (terminalReplayWins) {
      next = { ...next, pendingReplyText: '', liveReplyText: '' };
    } else {
      next = event.hydration.context;
      if (event.hydration.armTts && !side.some((item) => item.kind === 'armTts')) {
        side.push({ kind: 'armTts' });
      }
    }
    if ((!event.hydration.armTts || terminalReplayWins) && next.state === 'idle' && !next.pendingReplyText) {
      side = side.filter((item) => item.kind !== 'armTts');
    }
  }
  return { next, side };
}

function isTerminalReplayEvent(event: DrivingReplayEvent): boolean {
  return event.type === 'tts.done'
    || event.type === 'tts.error'
    || event.type === 'reply.error'
    || event.type === 'stt.error';
}

File: android/app/src/main/java/app/clawkietalkie/voice/DrivingReducer.kt
package app.clawkietalkie.voice

// Pure state reducer for the Driving screen. Mirror of the web client's
// `src/voice/drivingReducer.ts` — exposed independently of any UI framework
// so it can be unit-tested.
//
// States:
//   IDLE       → waiting for the user to tap start
//   RECORDING  → mic is open; daemon is recording audio for OpenClaw infer STT
//   THINKING   → mic closed; daemon running chat on the transcript
//   AI         → daemon running OpenClaw infer TTS; phone is playing audio

enum class DrivingState { IDLE, RECORDING, THINKING, AI }

data class DrivingContext(
    val state: DrivingState = DrivingState.IDLE,
    val lastUserText: String = "",
    val lastReplyText: String = "",
    // Reply text received before audio starts. Kept hidden until tts.start.
    val pendingReplyText: String = "",
    // Render label source for the AI caption.
    val liveReplyText: String = "",
    val error: String? = null,
)

val initialDrivingContext = DrivingContext()

sealed interface DrivingReplayEvent

sealed interface DrivingEvent {
    data class Tap(val currentTurnTranscribing: Boolean = false) : DrivingEvent, DrivingReplayEvent
    data object Silence : DrivingEvent, DrivingReplayEvent
    data class SttDone(val text: String) : DrivingEvent, DrivingReplayEvent
    data class SttError(val reason: String) : DrivingEvent, DrivingReplayEvent
    data class ReplyDone(val text: String) : DrivingEvent, DrivingReplayEvent
    data class ReplyError(val reason: String) : DrivingEvent, DrivingReplayEvent
    data class TtsStart(val text: String? = null) : DrivingEvent, DrivingReplayEvent
    data object TtsDone : DrivingEvent, DrivingReplayEvent
    data class TtsError(val reason: String) : DrivingEvent, DrivingReplayEvent
    data class SessionReplay(
        val events: List<DrivingReplayEvent>,
        val hydration: DrivingHydration? = null,
    ) : DrivingEvent

    data object SessionReset : DrivingEvent
}

data class DrivingHydration(
    val context: DrivingContext,
    val armTts: Boolean = false,
)

enum class DrivingSideEffect { START_MIC, STOP_MIC, CANCEL_MIC, ARM_TTS, STOP_TTS, CANCEL_REPLY }

data class Reduction(val next: DrivingContext, val side: List<DrivingSideEffect>)

fun reduce(ctx: DrivingContext, event: DrivingEvent): Reduction {
    if (event is DrivingEvent.SessionReplay) return reduceSessionReplay(ctx, event)
    if (event is DrivingEvent.SessionReset) return Reduction(initialDrivingContext, emptyList())

    return when (ctx.state) {
        DrivingState.IDLE -> when (event) {
            is DrivingEvent.Tap -> Reduction(
                ctx.copy(
                    state = DrivingState.RECORDING,
                    error = null,
                    pendingReplyText = "",
                    liveReplyText = "",
                ),
                listOf(DrivingSideEffect.START_MIC),
            )
            is DrivingEvent.TtsStart -> Reduction(
                if (!event.text.isNullOrEmpty()) {
                    ctx.copy(
                        state = DrivingState.AI,
                        lastReplyText = event.text,
                        liveReplyText = event.text,
                        error = null,
                    )
                } else {
                    ctx.copy(state = DrivingState.AI, error = null)
                },
                emptyList(),
            )
            else -> Reduction(ctx, emptyList())
        }

        DrivingState.RECORDING -> when (event) {
            is DrivingEvent.Tap -> Reduction(
                ctx.copy(state = DrivingState.THINKING),
                listOf(DrivingSideEffect.STOP_MIC),
            )
            is DrivingEvent.SttError -> Reduction(
                ctx.copy(state = DrivingState.IDLE, error = event.reason),
                listOf(DrivingSideEffect.CANCEL_MIC),
            )
            else -> Reduction(ctx, emptyList())
        }

        DrivingState.THINKING -> when (event) {
            is DrivingEvent.SttDone ->
                // Final transcript arrives while already in `thinking`. Don't
                // start anything new — daemon pipelines chat automatically.
                Reduction(ctx.copy(lastUserText = event.text), emptyList())
            is DrivingEvent.ReplyDone -> Reduction(
                ctx.copy(pendingReplyText = event.text, liveReplyText = ""),
                listOf(DrivingSideEffect.ARM_TTS),
            )
            is DrivingEvent.TtsStart -> {
                val replyText = ctx.pendingReplyText.ifEmpty { event.text ?: "" }
                if (replyText.isEmpty()) Reduction(ctx, emptyList())
                else Reduction(
                    ctx.copy(
                        state = DrivingState.AI,
                        lastReplyText = replyText,
                        liveReplyText = replyText,
                        pendingReplyText = "",
                    ),
                    emptyList(),
                )
            }
            is DrivingEvent.ReplyError -> Reduction(
                ctx.copy(
                    state = DrivingState.IDLE,
                    error = event.reason,
                    pendingReplyText = "",
                    liveReplyText = "",
                ),
                emptyList(),
            )
            is DrivingEvent.SttError -> Reduction(
                ctx.copy(
                    state = DrivingState.IDLE,
                    error = event.reason,
                    pendingReplyText = "",
                    liveReplyText = "",
                ),
                emptyList(),
            )
            is DrivingEvent.TtsDone -> Reduction(
                ctx.copy(state = DrivingState.IDLE, pendingReplyText = "", liveReplyText = ""),
                emptyList(),
            )
            is DrivingEvent.TtsError ->
                // Audio synthesis is non-fatal once the reply itself was
                // generated: promote the pending reply into the visible
                // "last AI turn" and surface a soft audio error.
                Reduction(
                    ctx.copy(
                        state = DrivingState.IDLE,
                        error = event.reason,
                        lastReplyText = ctx.pendingReplyText.ifEmpty { ctx.lastReplyText },
                        pendingReplyText = "",
                        liveReplyText = "",
                    ),
                    emptyList(),
                )
            is DrivingEvent.Tap -> {
                if (event.currentTurnTranscribing) Reduction(ctx, emptyList())
                else Reduction(
                    // Double-tap from thinking bails out of the turn after
                    // the authoritative STT final has arrived.
                    ctx.copy(state = DrivingState.IDLE, pendingReplyText = "", liveReplyText = ""),
                    listOf(DrivingSideEffect.CANCEL_REPLY),
                )
            }
            else -> Reduction(ctx, emptyList())
        }

        DrivingState.AI -> when (event) {
            is DrivingEvent.Tap, DrivingEvent.Silence -> Reduction(
                ctx.copy(state = DrivingState.IDLE),
                listOf(DrivingSideEffect.STOP_TTS),
            )
            DrivingEvent.TtsDone -> Reduction(ctx.copy(state = DrivingState.IDLE), emptyList())
            is DrivingEvent.TtsError -> Reduction(
                ctx.copy(state = DrivingState.IDLE, error = event.reason, liveReplyText = ""),
                emptyList(),
            )
            else -> Reduction(ctx, emptyList())
        }
    }
}

private fun isTerminalReplayEvent(event: DrivingReplayEvent): Boolean =
    event is DrivingEvent.TtsDone
        || event is DrivingEvent.TtsError
        || event is DrivingEvent.ReplyError
        || event is DrivingEvent.SttError

private fun reduceSessionReplay(ctx: DrivingContext, event: DrivingEvent.SessionReplay): Reduction {
    var next = ctx
    var side = mutableListOf<DrivingSideEffect>()
    val hasTerminalReplayEvent = event.events.any { isTerminalReplayEvent(it) }
    for (replayEvent in event.events) {
        val reduced = reduce(next, replayEvent as DrivingEvent)
        next = reduced.next
        side.addAll(reduced.side)
    }
    val hydration = event.hydration
    if (hydration != null) {
        val terminalReplayWins = hasTerminalReplayEvent &&
            next.state == DrivingState.IDLE &&
            hydration.context.state != DrivingState.IDLE
        if (terminalReplayWins) {
            next = next.copy(pendingReplyText = "", liveReplyText = "")
        } else {
            next = hydration.context
            if (hydration.armTts && DrivingSideEffect.ARM_TTS !in side) {
                side.add(DrivingSideEffect.ARM_TTS)
            }
        }
        if ((!hydration.armTts || terminalReplayWins) &&
            next.state == DrivingState.IDLE &&
            next.pendingReplyText.isEmpty()
        ) {
            side = side.filter { it != DrivingSideEffect.ARM_TTS }.toMutableList()
        }
    }
    return Reduction(next, side)
}
